Mage.Sets/src/mage/cards/e/EdgarMasterMachinist.java

```diff
@@ -0,0 +1,208 @@
+        super(AsThoughEffectType.CAST_FROM_NOT_OWN_HAND_ZONE, Duration.WhileOnBattlefield, Outcome.Benefit);
+        this.staticText = "Once during each of your turns, you may cast an artifact spell from your graveyard. "
+                + "If you cast a spell this way, that artifact enters tapped.";
+    }
+
+    private EdgarMasterMachinistCastFromGraveyardEffect(final EdgarMasterMachinistCastFromGraveyardEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public EdgarMasterMachinistCastFromGraveyardEffect copy() {
+        return new EdgarMasterMachinistCastFromGraveyardEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        return true;
+    }
+
+    @Override
+    public boolean applies(UUID objectId, Ability source, UUID affectedControllerId, Game game) {
+        throw new IllegalArgumentException("Wrong code usage: can't call applies method on empty affectedAbility");
+    }
+
+    @Override
+    public boolean applies(UUID objectId, Ability affectedAbility, Ability source, Game game, UUID playerId) {
+        Player controller = game.getPlayer(source.getControllerId());
+        Permanent sourcePermanent = source.getSourcePermanentIfItStillExists(game);
+        EdgarMasterMachinistWatcher watcher = game.getState().getWatcher(EdgarMasterMachinistWatcher.class);
+        Card cardToCast = game.getCard(objectId);
+        if (controller == null || sourcePermanent == null || watcher == null || cardToCast == null) {
+            return false;
+        }
+        if (!game.isActivePlayer(playerId) // only during your turn
+                || !source.isControlledBy(playerId) // only you may cast
+                || !Zone.GRAVEYARD.equals(game.getState().getZone(objectId)) // from graveyard
+                || !cardToCast.getOwnerId().equals(playerId) // only your graveyard
+                || !(affectedAbility instanceof SpellAbility) // characteristics to check
+                || watcher.abilityUsed(new MageObjectReference(sourcePermanent, game)) // once per turn
+        ) {
+            return false;
+        }
+        SpellAbility spellAbility = (SpellAbility) affectedAbility;
+        Card cardToCheck = spellAbility.getCharacteristics(game);
+        if (spellAbility.getManaCosts().isEmpty()) {
+            return false;
+        }
+        Set<MageIdentifier> allowedToBeCastNow = spellAbility.spellCanBeActivatedNow(playerId, game);
+        return allowedToBeCastNow.contains(MageIdentifier.Default)
+                && filter.match(cardToCheck, playerId, source, game);
+    }
+}
+
+class EdgarMasterMachinistEntersTappedEffect extends EntersBattlefieldEffect {
+
+    private final MageObjectReference mor;
+
+    EdgarMasterMachinistEntersTappedEffect(MageObjectReference mor) {
+        super(new TapSourceEffect(true));
+        this.mor = mor;
+        this.duration = Duration.EndOfTurn; // effect cleanup at end of turn if not consumed.
```

Comment:
Potential solution for precise cleanup: `Duration.OneUse` and `discard()` call on mor's outdate in `applies`